File: pbark/pbark/parser/controlflow/loop_control_parser.py
from __future__ import annotations

from pbark.lexer import TokenType
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords


class LoopControlParser:
    def __init__(self, parser) -> None:
        self._parser = parser

    def parse_break_line(self):
        if not self._line_starts_with(Keywords.BREAK_KEYWORDS):
            return None
        line = self._parser.peek().line
        self._parser.consume_until_line_end()
        return AstNode.Break(line)

    def parse_continue_line(self):
        if not self._line_starts_with(Keywords.CONTINUE_KEYWORDS):
            return None
        line = self._parser.peek().line
        self._parser.consume_until_line_end()
        return AstNode.Continue(line)

    def _line_starts_with(self, words: list[str]) -> bool:
        token = self._parser.peek()
        return token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in words

File: src/main/java/dev/klomptech/jbark/parser/controlflow/LoopControlParser.java
package dev.klomptech.jbark.parser.controlflow;

import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.AstNode;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.Parser;
import java.util.List;

public class LoopControlParser {

  private final Parser parser;

  public LoopControlParser(final Parser parser) {
    this.parser = parser;
  }

  // Example: "heel" on its own line inside a while/until/for-each body (break)
  public AstNode parseBreakLine() throws BarkError {
    if (!lineStartsWith(Keywords.BREAK_KEYWORDS)) {
      return null;
    }
    int line = parser.peek().line();
    parser.consumeUntilLineEnd();
    return new AstNode.Break(line);
  }

  // Example: "again" on its own line inside a loop body (continue)
  public AstNode parseContinueLine() throws BarkError {
    if (!lineStartsWith(Keywords.CONTINUE_KEYWORDS)) {
      return null;
    }
    int line = parser.peek().line();
    parser.consumeUntilLineEnd();
    return new AstNode.Continue(line);
  }

  // Break and continue must start the line, same as when/while/for
  private boolean lineStartsWith(final List<String> words) {
    Token token = parser.peek();
    return token.is(TokenType.IDENTIFIER) && words.contains(parser.normalise(token.value()));
  }
}
